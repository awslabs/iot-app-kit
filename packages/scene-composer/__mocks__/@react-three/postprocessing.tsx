module.exports = {
  EffectComposer: 'EffectComposer',
  Outline: 'Outline',
};
