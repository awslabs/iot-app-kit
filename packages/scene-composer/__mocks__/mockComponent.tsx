import React from 'react';

export default (componentName) => {
  // eslint-disable-next-line react/prop-types
  const component = ({ children, ...props }) => {
    const snapshotableProps = {};
    const renderProps = [];

    Object.entries(props).forEach(([key, value]) => {
      try {
        // Covers edge case where stringify fails due to circular object graphs
        if (typeof value === 'object') {
          if (React.isValidElement(value)) {
            renderProps[key] = value;
          } else if (key === 'style') {
            snapshotableProps[key] = value;
          } else {
            snapshotableProps[key] = JSON.stringify(value);
          }
        } else {
          snapshotableProps[key] = value;
        }
      } catch (e) {
        snapshotableProps[key] = value;
      }
    });

    return (
      <div data-mocked={componentName} {...snapshotableProps}>
        {Object.entries(renderProps).map(([key, value]) => (
          <div key={key}>{value}</div>
        ))}
        {children}
      </div>
    );
  };

   component.displayName = `Mock${componentName}`;
   return component
};
