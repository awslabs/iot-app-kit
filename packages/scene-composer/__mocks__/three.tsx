/**
 * Mock implementation of R3F (React Three Fiber) components with the string name for a clean snapshot test.
 */
module.exports = {
  ...jest.requireActual('three'),
  audioListener: 'audioListener',
  positionalAudio: 'positionalAudio',

  mesh: 'mesh',
  instancedMesh: 'instancedMesh',
  scene: 'scene',
  sprite: 'sprite',
  lOD: 'lOD',
  skinnedMesh: 'skinnedMesh',
  skeleton: 'skeleton',
  bone: 'bone',
  lineSegments: 'lineSegments',
  lineLoop: 'lineLoop',
  points: 'points',
  group: 'group',
  immediateRenderObject: 'immediateRenderObject',

  // cameras
  camera: 'camera',
  perspectiveCamera: 'perspectiveCamera',
  orthographicCamera: 'orthographicCamera',
  cubeCamera: 'cubeCamera',
  arrayCamera: 'arrayCamera',

  // geometry
  instancedBufferGeometry: 'instancedBufferGeometry',
  bufferGeometry: 'bufferGeometry',
  boxBufferGeometry: 'boxBufferGeometry',
  circleBufferGeometry: 'circleBufferGeometry',
  coneBufferGeometry: 'coneBufferGeometry',
  cylinderBufferGeometry: 'cylinderBufferGeometry',
  dodecahedronBufferGeometry: 'dodecahedronBufferGeometry',
  extrudeBufferGeometry: 'extrudeBufferGeometry',
  icosahedronBufferGeometry: 'icosahedronBufferGeometry',
  latheBufferGeometry: 'latheBufferGeometry',
  octahedronBufferGeometry: 'octahedronBufferGeometry',
  parametricBufferGeometry: 'parametricBufferGeometry',
  planeBufferGeometry: 'planeBufferGeometry',
  polyhedronBufferGeometry: 'polyhedronBufferGeometry',
  ringBufferGeometry: 'ringBufferGeometry',
  shapeBufferGeometry: 'shapeBufferGeometry',
  sphereBufferGeometry: 'sphereBufferGeometry',
  tetrahedronBufferGeometry: 'tetrahedronBufferGeometry',
  textBufferGeometry: 'textBufferGeometry',
  torusBufferGeometry: 'torusBufferGeometry',
  torusKnotBufferGeometry: 'torusKnotBufferGeometry',
  tubeBufferGeometry: 'tubeBufferGeometry',
  wireframeGeometry: 'wireframeGeometry',
  parametricGeometry: 'parametricGeometry',
  tetrahedronGeometry: 'tetrahedronGeometry',
  octahedronGeometry: 'octahedronGeometry',
  icosahedronGeometry: 'icosahedronGeometry',
  dodecahedronGeometry: 'dodecahedronGeometry',
  polyhedronGeometry: 'polyhedronGeometry',
  tubeGeometry: 'tubeGeometry',
  torusKnotGeometry: 'torusKnotGeometry',
  torusGeometry: 'torusGeometry',
  textGeometry: 'textGeometry',
  sphereGeometry: 'sphereGeometry',
  ringGeometry: 'ringGeometry',
  planeGeometry: 'planeGeometry',
  latheGeometry: 'latheGeometry',
  shapeGeometry: 'shapeGeometry',
  extrudeGeometry: 'extrudeGeometry',
  edgesGeometry: 'edgesGeometry',
  coneGeometry: 'coneGeometry',
  cylinderGeometry: 'cylinderGeometry',
  circleGeometry: 'circleGeometry',
  boxGeometry: 'boxGeometry',

  // materials
  material: 'material',
  shadowMaterial: 'shadowMaterial',
  spriteMaterial: 'spriteMaterial',
  rawShaderMaterial: 'rawShaderMaterial',
  shaderMaterial: 'shaderMaterial',
  pointsMaterial: 'pointsMaterial',
  meshPhysicalMaterial: 'meshPhysicalMaterial',
  meshStandardMaterial: 'meshStandardMaterial',
  meshPhongMaterial: 'meshPhongMaterial',
  meshToonMaterial: 'meshToonMaterial',
  meshNormalMaterial: 'meshNormalMaterial',
  meshLambertMaterial: 'meshLambertMaterial',
  meshDepthMaterial: 'meshDepthMaterial',
  meshDistanceMaterial: 'meshDistanceMaterial',
  meshBasicMaterial: 'meshBasicMaterial',
  meshMatcapMaterial: 'meshMatcapMaterial',
  lineDashedMaterial: 'lineDashedMaterial',
  lineBasicMaterial: 'lineBasicMaterial',

  // primitive
  primitive: 'primitive',

  // lights and other
  light: 'light',
  spotLightShadow: 'spotLightShadow',
  spotLight: 'spotLight',
  pointLight: 'pointLight',
  rectAreaLight: 'rectAreaLight',
  hemisphereLight: 'hemisphereLight',
  directionalLightShadow: 'directionalLightShadow',
  directionalLight: 'directionalLight',
  ambientLight: 'ambientLight',
  lightShadow: 'lightShadow',
  ambientLightProbe: 'ambientLightProbe',
  hemisphereLightProbe: 'hemisphereLightProbe',
  lightProbe: 'lightProbe',

  // helpers
  spotLightHelper: 'spotLightHelper',
  skeletonHelper: 'skeletonHelper',
  pointLightHelper: 'pointLightHelper',
  hemisphereLightHelper: 'hemisphereLightHelper',
  gridHelper: 'gridHelper',
  polarGridHelper: 'polarGridHelper',
  directionalLightHelper: 'directionalLightHelper',
  cameraHelper: 'cameraHelper',
  boxHelper: 'boxHelper',
  box3Helper: 'box3Helper',
  planeHelper: 'planeHelper',
  arrowHelper: 'arrowHelper',
  axesHelper: 'axesHelper',

  // textures
  texture: 'texture',
  videoTexture: 'videoTexture',
  dataTexture: 'dataTexture',
  dataTexture3D: 'dataTexture3D',
  compressedTexture: 'compressedTexture',
  cubeTexture: 'cubeTexture',
  canvasTexture: 'canvasTexture',
  depthTexture: 'depthTexture',

  // misc
  raycaster: 'raycaster',
  vector2: 'vector2',
  vector3: 'vector3',
  vector4: 'vector4',
  euler: 'euler',
  matrix3: 'matrix3',
  matrix4: 'matrix4',
  quaternion: 'quaternion',
  bufferAttribute: 'bufferAttribute',
  instancedBufferAttribute: 'instancedBufferAttribute',
  color: 'color',
  fog: 'fog',
  fogExp2: 'fogExp2',
  shape: 'shape',
};
