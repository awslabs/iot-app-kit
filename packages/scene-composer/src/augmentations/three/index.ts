export * from './Anchor';
export * from './Viewpoint';
