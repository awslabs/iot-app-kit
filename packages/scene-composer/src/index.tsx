export * from './interfaces';

export * from './errors';

export * from './SceneComposer';

// Augmentation related interfaces
export * from './augmentations/three';

export * from './GlobalSettings';

export * from './utils/dataFrameLabelUtils';

export * from './utils/dataBindingTemplateUtils';

export * from './constants';
