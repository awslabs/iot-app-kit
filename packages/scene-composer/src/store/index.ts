export * from './StoreOperations';
export * from './Store';
