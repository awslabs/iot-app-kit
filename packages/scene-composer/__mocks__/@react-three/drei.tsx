import mockComponent from '../mockComponent';

const dreiRaw: any = jest.createMockFromModule('@react-three/drei');

const drei = Object.keys(dreiRaw).reduce((acc, comp) => {
  if (!comp.startsWith('_')) {
    acc[comp] = mockComponent(comp);
  } else {
    acc[comp] = dreiRaw[comp];
  }
  return acc;
}, {});

module.exports = {
  ...drei,
};
