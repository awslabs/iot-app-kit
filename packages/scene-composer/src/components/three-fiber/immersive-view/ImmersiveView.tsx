import React, { ReactNode, useContext, useEffect, useState, useRef, useMemo } from 'react';
import * as THREE from 'three';
import { useFrame, useLoader, useThree } from '@react-three/fiber';
import { Mesh } from 'three';
import { IntlProvider } from 'react-intl';
import { DepthOfField, EffectComposer, Vignette } from '@react-three/postprocessing';

import useLogger from '../../../logger/react-logger/hooks/useLogger';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { IViewpointComponentInternal, useEditorState, useSceneDocument } from '../../../store';
import { KnownComponentType } from '../../../interfaces';
import { findComponentByType } from '../../../utils/nodeUtils';
import { LoadingProgress } from '../LoadingProgress';
import usePreloadSkyboxImages from '../../../hooks/usePreloadSkyboxImages';
import { getGlobalSettings } from '../../../common/GlobalSettings';

import { generateCubeMapSkyboxTexture, generateSixSidedSkyboxTexture } from './ImmersiveViewUtils';

export const generateEquirectangularSkyboxTexture = (image: string, renderer: THREE.WebGLRenderer): THREE.Texture => {
  // TODO: Not working yet, need to fix, and then update to use TwinMakerTextureLoader
  const texture: THREE.Texture = useLoader(THREE.TextureLoader, image);
  const renderTarget = new THREE.WebGLCubeRenderTarget(texture.image.height);
  renderTarget.fromEquirectangularTexture(renderer, texture);
  return renderTarget.texture;
};

export const createMaterialNodes = (textures: THREE.Texture[]): ReactNode[] => {
  const materials: ReactNode[] = [];

  textures.forEach((texture, index) => {
    materials.push(<meshBasicMaterial key={index} map={texture} attachArray={'material'} side={THREE.BackSide} />);
  });

  return materials;
};

export const createMaterial = (texture: THREE.Texture): ReactNode => {
  // TODO: In the case of using an actual box geometry vs the scene background we have an issue loading
  //  equirectangular images. The outside of the box is mapped correctly but looking at it
  //  from the inside has the x, y, and z images flipped
  //  https://sim.amazon.com/issues/IOTROCI-5883
  return <meshBasicMaterial envMap={texture} side={THREE.BackSide} />;
};

export function ImmersiveView() {
  const log = useLogger('ImmersiveView');
  const sceneComposerId = useContext(sceneComposerIdContext);
  const { cameraControlsType, selectedViewpointNodeRef, getObject3DBySceneNodeRef, isInViewpointTransition } =
    useEditorState(sceneComposerId);
  const { getSceneNodeByRef } = useSceneDocument(sceneComposerId);

  const { gl, camera } = useThree();
  const renderer = gl;
  const node = getSceneNodeByRef(selectedViewpointNodeRef);
  const viewpointComponent = useMemo(() => {
    return findComponentByType(node, KnownComponentType.Viewpoint) as IViewpointComponentInternal;
  }, [node, selectedViewpointNodeRef]);

  const [textureFormat, setTextureFormat] = useState<string>();
  const [textures, setTextures] = useState<THREE.Texture[] | null>(null);
  const [viewpointPosition, setViewpointPosition] = useState<THREE.Vector3>();
  const [viewpointRotation, setViewpointRotation] = useState<THREE.Vector3>();
  const [isReady, setIsReady] = useState<boolean>(false);

  const [shouldUpdateUV, setShouldUpdateUV] = useState<boolean>(true);

  const meshRef = useRef<Mesh>();

  usePreloadSkyboxImages();

  useEffect(() => {
    if (cameraControlsType === 'immersive' && viewpointComponent) {
      // Create Skybox texture and set to scene background.
      setTextureFormat(viewpointComponent.skyboxImageFormat);
      setIsReady(viewpointComponent?.skyboxImages.some((image) => THREE.Cache.get(image)));

      // Update position
      const viewpointObject = getObject3DBySceneNodeRef(node?.ref);

      const cameraPosition = new THREE.Vector3(
        viewpointComponent.cameraPosition[0],
        viewpointComponent.cameraPosition[1],
        viewpointComponent.cameraPosition[2],
      );

      const cameraRotation = viewpointComponent.cameraRotation
        ? new THREE.Vector3(
            viewpointComponent.cameraRotation[0],
            viewpointComponent.cameraRotation[1],
            viewpointComponent.cameraRotation[2],
          )
        : new THREE.Vector3(0, 0, 0);

      const position = cameraPosition;
      const rotation = cameraRotation;
      let parent: THREE.Object3D | null = viewpointObject?.parent || null;
      while (parent) {
        position.add(parent.position);
        rotation.add(parent.rotation.toVector3());
        parent = parent.parent;
      }

      setViewpointPosition(position);
      setViewpointRotation(rotation);

      switch (viewpointComponent.skyboxImageFormat) {
        case 'SixSided':
          generateSixSidedSkyboxTexture(viewpointComponent.skyboxImages, setTextures).finally(() => {
            setIsReady(true);
          });
          break;
        case 'CubeMap':
          generateCubeMapSkyboxTexture(viewpointComponent.skyboxImages[0], setTextures).finally(() => {
            setIsReady(true);
          });
          break;
        case 'Equirectangular':
          setTextures([generateEquirectangularSkyboxTexture(viewpointComponent.skyboxImages[0], renderer)]);
          break;
        default:
          setTextures(null);
          break;
      }
    } else {
      setTextures(null);
    }
  }, [viewpointComponent, cameraControlsType]);

  useEffect(() => {
    setShouldUpdateUV(true);
  }, [cameraControlsType, selectedViewpointNodeRef]);

  useFrame(() => {
    // recalculate uv mapping to solve seam on the edge issue
    // https://sim.amazon.com/IOTROCI-6327
    const offsetParameter = 0.005;
    if (meshRef.current?.geometry && shouldUpdateUV) {
      setShouldUpdateUV(false);
      const oldUvMapping = meshRef.current.geometry.attributes.uv;
      const newUvMapping: number[] = [];
      for (let i = 0; i < oldUvMapping.array.length; i++) {
        const offset =
          oldUvMapping.array[i] === 0
            ? offsetParameter
            : oldUvMapping.array[i] === 1
            ? 1 - offsetParameter
            : oldUvMapping.array[i];

        newUvMapping.push(offset);
      }

      meshRef.current.geometry.setAttribute('uv', new THREE.Float32BufferAttribute(newUvMapping, 2));
    }
  });

  const size = camera.far;

  const render = () => {
    if (cameraControlsType === 'immersive') {
      if (!isReady) {
        return (
          <IntlProvider locale={getGlobalSettings().locale}>
            <LoadingProgress />
          </IntlProvider>
        );
      } else if (isReady && textures) {
        return (
          <>
            <mesh
              ref={meshRef}
              type={'Skybox'}
              renderOrder={-1}
              position={viewpointPosition ? [viewpointPosition.x, viewpointPosition.y, viewpointPosition.z] : [0, 0, 0]}
              rotation={viewpointRotation ? [viewpointRotation.x, viewpointRotation.y, viewpointRotation.z] : [0, 0, 0]}
            >
              <boxGeometry args={[size, size, size]} />
              {textureFormat === 'Equirectangular' ? createMaterial(textures[0]) : createMaterialNodes(textures)}
            </mesh>
            {isInViewpointTransition && (
              <EffectComposer>
                <DepthOfField focusDistance={1} focalLength={0.02} bokehScale={7} height={480} />
                <Vignette offset={0.5} />
              </EffectComposer>
            )}
          </>
        );
      }
    }

    return null;
  };

  return <>{render()}</>;
}
