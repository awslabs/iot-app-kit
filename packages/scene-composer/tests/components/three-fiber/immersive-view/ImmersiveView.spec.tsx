/* eslint-disable import/first */
/* eslint-disable import/order */
import * as THREE from 'three';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { useThree } from '@react-three/fiber';

import { mockR3F } from '../../../__mocks__/MockR3F';
mockR3F();

import { ImmersiveView } from '../../../../src/components/three-fiber/immersive-view/ImmersiveView';
import { useStore } from '../../../../src/store';
import { KnownComponentType } from '../../../../src';
import Mock = jest.Mock;

jest.mock('three', () => {
  const originalModule = jest.requireActual('three');
  class MockTexture {}
  return {
    ...originalModule,
    Texture: MockTexture,
  };
});

jest.mock('../../../../src/components/three-fiber/LoadingProgress', () => ({
  LoadingProgress: 'LoadingProgress',
}));

jest.mock('../../../../src/hooks/usePreloadSkyboxImages', () => jest.fn());

jest.mock('@react-three/fiber', () => {
  const originalModule = jest.requireActual('@react-three/fiber');
  return {
    ...originalModule,
    useThree: jest.fn(),
    useFrame: jest.fn().mockImplementation((callback) => callback()),
  };
});

jest.mock('../../../../src/components/three-fiber/immersive-view/ImmersiveViewUtils', () => ({
  areAllImagesInCache: jest.fn(),
  generateSixSidedSkyboxTexture: jest.fn().mockImplementation((images, setter) => {
    setter([new THREE.Texture()]);
    return new Promise(jest.fn());
  }),
}));

describe('ImmersiveView', () => {
  const viewpointObject = new THREE.Object3D();
  const boxGeometry = new THREE.BoxGeometry(10, 10, 10);
  const material = new THREE.MeshBasicMaterial();
  const mesh = new THREE.Mesh(boxGeometry, material);
  mesh.position.set(1, 1, 1);
  mesh.add(viewpointObject);
  const scene = new THREE.Scene();
  scene.add(mesh);

  const mockGetSceneNodeByRef = jest.fn();
  const mockSetCameraControlsType = jest.fn();
  const mockGetObject3DBySceneNodeRef = jest.fn().mockReturnValue(viewpointObject);

  const sixSidedViewpointNode = {
    ref: 'selected-viewpoint-ref',
    components: [
      {
        type: KnownComponentType.Viewpoint,
        skyboxImageFormat: 'SixSided',
        skyboxImages: ['px.png', 'nx.png', 'py.png', 'ny.png', 'pz.png', 'nz.png'],
        cameraPosition: [0, 1.5, 0],
        visibleObjectIds: [],
      },
    ],
  };

  const cubeMapViewpointNode = {
    ref: 'selected-viewpoint-ref',
    components: [
      {
        type: KnownComponentType.Viewpoint,
        skyboxImageFormat: 'CubeMap',
        skyboxImages: ['cubemap.png'],
        cameraPosition: [0, 1.5, 0],
        visibleObjectIds: [],
      },
    ],
  };

  const equirectangularViewpointNode = {
    ref: 'selected-viewpoint-ref',
    components: [
      {
        type: KnownComponentType.Viewpoint,
        skyboxImageFormat: 'Equirectangular',
        skyboxImages: ['equirectangular.png'],
        cameraPosition: [0, 1.5, 0],
        visibleObjectIds: [],
      },
    ],
  };

  const baseState: any = {
    getSceneNodeByRef: mockGetSceneNodeByRef,
    setCameraControlsType: mockSetCameraControlsType,
    getObject3DBySceneNodeRef: mockGetObject3DBySceneNodeRef,
    selectedSceneNodeRef: 'selected-ref',
    selectedViewpointNodeRef: ' selected-viewpoint-ref',
    cameraControlsType: 'immersive',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    useStore('default').setState(baseState);
    (useThree as unknown as Mock).mockReturnValue({
      renderer: 'mockRenderer',
      camera: new THREE.PerspectiveCamera(),
    });
  });

  it('should render Loading state correctly', () => {
    useStore('default').setState(baseState);
    mockGetSceneNodeByRef.mockReturnValue(sixSidedViewpointNode);

    const container = renderer.create(<ImmersiveView />);
    expect(container).toMatchSnapshot();
  });

  it('should render six sided Viewpoint correctly', () => {
    mockGetSceneNodeByRef.mockReturnValue(sixSidedViewpointNode);
    jest.spyOn(THREE.Cache, 'get').mockReturnValue(new THREE.Texture());

    let container;
    act(() => {
      container = renderer.create(<ImmersiveView />);
    });
    expect(container).toMatchSnapshot();
  });

  it('should set uv mapping as expected.', () => {
    const boxGeometry = new THREE.BoxGeometry(1, 1, 1);
    jest.spyOn(React, 'useRef').mockImplementation(() => {
      return { current: { geometry: boxGeometry } };
    });

    act(() => {
      renderer.create(<ImmersiveView />);
    });

    expect(Math.abs(0.005 - boxGeometry.attributes.uv.array[0])).toBeLessThan(0.00001);
  });
});
