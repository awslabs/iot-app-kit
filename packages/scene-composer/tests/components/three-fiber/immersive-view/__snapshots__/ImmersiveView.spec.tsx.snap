// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImmersiveView should render Loading state correctly 1`] = `<LoadingProgress />`;

exports[`ImmersiveView should render six sided Viewpoint correctly 1`] = `
<mesh
  position={
    Array [
      1,
      2.5,
      1,
    ]
  }
  renderOrder={-1}
  rotation={
    Array [
      0,
      0,
      0,
    ]
  }
  type="Skybox"
>
  <boxGeometry
    args={
      Array [
        2000,
        2000,
        2000,
      ]
    }
  />
  <meshBasicMaterial
    attachArray="material"
    map={MockTexture {}}
    side={1}
  />
</mesh>
`;
